### PR TITLE
Function featurizer

### PR DESCRIPTION
**Pull request by msultan**
Featurizer capable of hopefully taking any arbitrary function and mapping it to the input set of trajectories. I wrote one test though i dont know how this is gonna get put into the command line.

### CODE REVIEW

**Comment by mpharrigan** [comment]
Let's make the api *only* accept a function (/callable).

If we want to expose this on the command line, then we can include a thing in the command-line parser to load from a pickle file

File: msmbuilder/featurizer/featurizer.py [hunk 1]
```diff
@@ -307,6 +309,56 @@ def partial_transform(self, traj):
...
+    ----------
+    function : instatiation of the function
+
+    """
+
+    def __init__(self, function, **kwargs):
+        if hasattr(function, '__call__'):
+            self.function = function
+            self.args = {}
+            self.args.update(kwargs)
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: why not `self.args = kwargs`?

File: msmbuilder/featurizer/featurizer.py [hunk 2]
```diff
@@ -307,6 +309,56 @@ def partial_transform(self, traj):
...
+        elif os.path.isfile(function):
+            self.function = verboseload(function)
+        else:
+            raise Exception("Sorry but we "
+                            "couldn't use the provided"
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: Missing a space

File: msmbuilder/featurizer/featurizer.py [hunk 3]
```diff
@@ -307,6 +309,56 @@ def partial_transform(self, traj):
...
+                            "function.")
+
+    def partial_transform(self, traj):
+        """Featurize an MD trajectory into a vector space via
+        the input function
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: First line of docstring should be one complete sentence (aka, don't wrap here)

File: msmbuilder/tests/test_featurizer.py [hunk 4]
```diff
@@ -48,6 +50,30 @@ def test_SubsetAtomPairs2():
...
+    trj0 = trajectories[0][0]
+    #use the dihedral to compute phi for ala
+
+
+    atom_ind = [[4, 6, 8,14]]
+    func = compute_dihedrals
+    f = FunctionFeaturizer(func, indices=atom_ind)
+    res1 = f.transform([trj0])
+
+    def funcception(trj):
```
Comment (mpharrigan) [inline msmbuilder/tests/test_featurizer.py]: lol

File: msmbuilder/tests/test_featurizer.py [hunk 5]
```diff
@@ -48,6 +50,30 @@ def test_SubsetAtomPairs2():
...
+        return compute_phi(trj)[1]
+
+    f = FunctionFeaturizer(funcception)
+    res2 = f.transform([trj0])
+
+    f3 = DihedralFeaturizer(['phi'], sincos=False)
+    res3 = f3.transform([trj0])
+
+    assert res1==res2==res3
```
Comment (mpharrigan) [inline msmbuilder/tests/test_featurizer.py]: Does this work? Shouldn't you use `np.testing.assert_array_equal` or something like that to check elementwise equality

File: msmbuilder/tests/test_featurizer.py [hunk 6]
```diff
@@ -48,6 +50,30 @@ def test_SubsetAtomPairs2():
...
+    return
```
Comment (mpharrigan) [inline msmbuilder/tests/test_featurizer.py]: no return

File: msmbuilder/featurizer/featurizer.py [hunk 7]
```diff
@@ -307,6 +307,73 @@ def partial_transform(self, traj):
...
+            A featurized trajectory is a 2D array of shape
+            `(length_of_trajectory x n_features)` where each `features[i]`
+            vector is computed by applying the featurization function
+            to the `i`th snapshot of the input trajectory.
+
+        See Also
+        --------
+        transform : simultaneously featurize a collection of MD trajectories
+
+         Notes:
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: - remove space before this
- no colon
- remove extra underbar

File: msmbuilder/featurizer/featurizer.py [hunk 8]
```diff
@@ -307,6 +307,73 @@ def partial_transform(self, traj):
...
+class FunctionFeaturizer(Featurizer):
+    """Featurizer based on arbitrary functions.
+
+    This featurizer transforms a dataset containing MD trajectories into
+    a vector dataset by representing each frame in each of the MD trajectories
+    by a vector the output of the function.
+
+    Parameters
+    ----------
+    function : Instantiation of the function
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: use the numpydoc format
```
name : type
    description
```

File: msmbuilder/featurizer/featurizer.py [hunk 9]
```diff
@@ -307,6 +307,73 @@ def partial_transform(self, traj):
...
+    *args : arguments to pass to the function MINUS the trajectory keyword
+    **kwargs: key word arguments to pass to the function MINUS the trajectory
+    keyword
+
+    UseCase:
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: This is not a valid section

Comment (msultan) [inline msmbuilder/featurizer/featurizer.py]: Should i move it to the notes? I feel having an example here is going to be useful.

Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: I think there's something like "example". check https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

File: msmbuilder/featurizer/featurizer.py [hunk 10]
```diff
@@ -307,6 +307,75 @@ def partial_transform(self, traj):
...
+    Parameters
+    ----------
+    function : function
+        Instantiation of the function. The function should accept
+        a mdtraj.Trajectory object as the first argument.
+    *args : list
+        arguments to pass to the function MINUS the trajectory keyword
+    **kwargs: kwargs
+        key word arguments to pass to the function MINUS the trajectory
+    keyword
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: indent

File: msmbuilder/featurizer/featurizer.py [hunk 11]
```diff
@@ -307,6 +307,77 @@ def partial_transform(self, traj):
...
+
+    Examples
+    --------
+    >>> function = compute_dihedrals
+    >>> f = FunctionFeaturizer(function, indices=[[0,1,2,3]])
+    >>> results = f.transform(dataset)
+    """
+
+    def __init__(self, function, *args, **kwargs):
+        if hasattr(function, '__call__'):
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: I think using the builtin callable covers more use-cases here

https://docs.python.org/3.4/library/functions.html#callable

**Comment by mpharrigan** [comment]
For the api, I think it would be more robust for the `__init__` method to take a dictionary of keyword arguments rather than `**kwargs`

```python
FunctionFeaturizer(func=mdtraj.compute_dihedrals, func_args={'indices': [[0,1,2,3]]})
```

I can see the current scheme messing up (1) the command line generation and (2) sklearn pipelines, both of which rely on introspection and don't expect `*args` or `**kwargs`

File: msmbuilder/featurizer/featurizer.py (hunk 11 above)
Comment (msultan) [inline msmbuilder/featurizer/featurizer.py]: ohh, didnt know about that.

**Comment by msultan** [comment]
> 
```python
FunctionFeaturizer(func=mdtraj.compute_dihedrals, func_args={'indices': [[0,1,2,3]]})
```
alright, i will switch it.

**Comment by msultan** [comment]
hmm weird, tests seem to be failing for at random places.

File: msmbuilder/featurizer/featurizer.py [hunk 12]
```diff
@@ -307,6 +307,75 @@ def partial_transform(self, traj):
...
+        if callable(function):
+            self.function = function
+            self.func_args = func_args
+        else:
+            raise Exception("Sorry but we "
+                            "couldn't use the provided "
+                            "function.")
+
+    def partial_transform(self, traj):
+        """Featurize a MD trajectory into a vector by
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: Can you make this a one-line summary?

File: msmbuilder/featurizer/featurizer.py [hunk 13]
```diff
@@ -307,6 +307,75 @@ def partial_transform(self, traj):
...
+
+        Notes
+        -----
+        This method assumes that the function takes in the trajectory object
+        as the first argument.
+
+        """
+        x = []
+        x.append(self.function(traj,  **self.func_args))
+        return np.hstack(x)
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: These three lines are redundant. Just return `self.function(traj, **self.func_args)`

File: msmbuilder/featurizer/featurizer.py [hunk 14]
```diff
@@ -307,6 +307,75 @@ def partial_transform(self, traj):
...
+    >>> f = FunctionFeaturizer(function, func_args={indices: [[0,1,2,3]]})
+    >>> results = f.transform(dataset)
+    """
+
+    def __init__(self, function, func_args={}):
+        if callable(function):
+            self.function = function
+            self.func_args = func_args
+        else:
+            raise Exception("Sorry but we "
```
Comment (mpharrigan) [inline msmbuilder/featurizer/featurizer.py]: This should probably be `ValueError` and maybe include the fact that the object isn't callable in the error message so the user knows what's wrong

Comment (cxhernandez) [inline msmbuilder/featurizer/featurizer.py]: +1

Comment (msultan) [inline msmbuilder/featurizer/featurizer.py]: that makes more sense. Making the changes now.

**Comment by mpharrigan** [comment]
:+1:

**Comment by msultan** [comment]
aww i wanted to merge this

**Pull request merged**